File: problems/0006-zigzag-conversion/solution.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1) { return s; }
        
        
        vector<string> rows;
        string solution;
        for (int i = 0; i < numRows; i++) {
            rows.push_back("");
        }
        
        int curChar = 0;
        int row = 0;
        int increment = 1;
        
        while (curChar < s.size()) {
            rows[row] += s[curChar];
            curChar++;
            if (row == numRows - 1) { increment = -1; }
            if (row == 0) { increment = 1; }
            row += increment;
        }
        
        for (int i = 0; i < numRows; i++) {
            solution += rows[i];
        }
        return solution;
    }
};

File: problems/0006-zigzag-conversion/solution.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
            
        rowIdx = 0
        lastRowIdx = numRows - 1
        goingDown = True
        rows = [""] * numRows

        for i, c in enumerate(s):
            rows[rowIdx] += c
            if goingDown:
                rowIdx += 1
            else:
                rowIdx -= 1

            if rowIdx == lastRowIdx:
                goingDown = False
            if rowIdx == 0:
                goingDown = True

        ans = ""
        for s in rows:
            ans += s
        
        return ans


